File: python/src/a2a_utils/storage/path_safety.py
"""Path safety utilities for filesystem-backed storage."""

from pathlib import Path
from .identifiers import IDENTIFIER_RE, assert_identifier

SAFE_STORAGE_ID_RE = IDENTIFIER_RE


def assert_safe_storage_id(name: str, value: str) -> None:
    """Reject IDs that could escape or corrupt the storage layout."""

    assert_identifier(name, value)


def assert_within_root(root: Path, target: Path) -> None:
    """Ensure a target path stays within the resolved storage root."""

    resolved_root = root.resolve()
    resolved_target = target.resolve(strict=False)

    try:
        resolved_target.relative_to(resolved_root)
    except ValueError as exc:
        raise ValueError(f"Path escapes storage root: '{target}'") from exc


def safe_join(root: Path, *segments: str) -> Path:
    """Join path segments while enforcing the storage root boundary."""

    target = root.joinpath(*segments)
    assert_within_root(root, target)
    return target

File: javascript/src/storage/path-safety.ts
// SPDX-FileCopyrightText: 2025-present A2A Net <hello@a2anet.com>
//
// SPDX-License-Identifier: Apache-2.0

import * as path from "node:path";
import { UUID_RE, assertUuid } from "./identifiers.js";

export const SAFE_STORAGE_ID_RE = UUID_RE;

export function assertSafeStorageId(name: string, value: string): void {
    assertUuid(name, value);
}

export function assertWithinRoot(root: string, target: string): void {
    const resolvedRoot = path.resolve(root);
    const resolvedTarget = path.resolve(target);
    const relative = path.relative(resolvedRoot, resolvedTarget);

    if (relative.startsWith("..") || path.isAbsolute(relative)) {
        throw new Error(`Path escapes storage root: '${target}'`);
    }
}

export function safeJoin(root: string, ...segments: string[]): string {
    const target = path.join(root, ...segments);
    assertWithinRoot(root, target);
    return target;
}
